### PR TITLE
Fix regressions and improve the insert link search field and suggestions list

### LINKED ISSUE #10820: Links suggestions: suggestions list doesn't scroll when using a keyboard
**Describe the bug**

When there are many suggested links (more than 10) and using a keyboard to highlight the suggested links, the list doesn't scroll.

**To Reproduce**
- create a post, add a block, select some text and start inserting a link
- using only the keyboard type something in the search field so you have more than 10 results
- only ~10 results are visible in the list (note: there are some CSS issues on current master)
- in my case, see screenshot below, I have 20 results and only 10 are initially visible:

screen shot 2018-10-20 at 00 32 52

- now, use the keyboard down or up arrow keys to highlight the results in the list
- the "highlighting" moves through the list
- but when on the 11th result, the list doesn't scroll
- keep pressing arrow down and apparently nothing happens (while actually it is highlighting the results that are not visible)
- at the 21st key press, finally the first result on top is highlighted again

This way, it's impossible to actually _see_ the complete list of results.

Note that while focus should stay in the search field, the list should scroll when pressing down / up arrows. In the Classic Editor this works like a charme.

### PR DESCRIPTION
**Pull request by afercia**
Fixes #10820 

Additionally fixes a few CSS regressions and improves the interaction with the insert link search field and suggestions list.

First commit:
- regression: makes the list of suggestions scrollable again when using a keyboard and the list is longer than the max-height (it worked in 2.9.0)
- regression: fixes the input padding after https://github.com/WordPress/gutenberg/commit/38378c88a0fe73c6413dd13c16a3830a433ae0bf which broke the usage of the variable in the padding calculation
- regression: fixes the spinner position for the same reason

Worth noting:
- I couldn't reproduce the original issue mentioned in #8646 because there's now a 1 pixel padding that makes the focus style visible so I just reverted the variable to `8px`
- re: the padding on the right: seems to me there's nothing to display in that space so I'm not sure why the padding was meant to be `44px`; changed to `8px`

Second commit:
- makes the `TAB` key populate the search field with the highlighted link URL
- this allows keyboard users to select the link (not submit yet) and tab to the Link Settings button; matches the current interaction in the Classic Editor
- adds a speak message to announce the link has been selected, matches the current interaction in the Classic Editor
- when clicking on a suggested link, moves focus to the search field: matches the current interaction in the Classic Editor and clarifies a further action (submit) is needed
- changes the `Link added.` message to `Link inserted` to match the Classic Editor message

Third commit:
- ~~regression: fixes the position of the suggestions popover after #10495~~ 
- the suggestions popover is now nested within another popover (the `URLPopover` introduced in #10495)
- this broke the position calculation, on current master the position looks like in the screenshot below:

screen shot 2018-10-20 at 00 32 52

Basically, this happens because the "Link Settings" button is now placed outside of the first popover:
screen shot 2018-10-20 at 17 12 26

Thus, the calculation of the suggestions popover is based on the search field _form_ without the settings button. I can think of two ways to fix this:
- either move back the "Link Settings" button within the form (not sure if this is possible / desirable)
- or use a custom `getAnchorRect` function to get the grand parent node

I've opted for the latter, even if a better solution would probably be making the popover component handle these nesting cases. I'd appreciate some eyes form @aduth and @talldan. If there's the need for adjustments / refinements / improvements glad if you can take this PR and make it better.

- adds 2 pixels to the suggestions list width to match the search field popover width (it has now an additional 1 pixel padding)
- removes `width: 100%;` from the search field because it made the following "Apply" button smaller than it's supposed to be (the correct width is 36 pixels)
- removes a stray `list-style: none;` applied on a div

Re: the popover position when a popover is nested within another popover, I guess this relates also to #8468.

### CODE REVIEW

File: packages/editor/src/components/url-input/index.js [hunk 1]
```diff
@@ -22,13 +22,23 @@ import { addQueryArgs } from '@wordpress/url';
 // as being considered from the input.
 const stopEventPropagation = ( event ) => event.stopPropagation();
 
+function getAnchorRect( anchor ) {
+	/*
+	 * The default getAnchorRect() gets the parent node to calculate the popover
+	 * position. As the popover with the list of suggestions is now nested within
+	 * another popover, we need to get the gran parent node.
+	 */
+	return anchor.parentNode.parentNode.getBoundingClientRect();
```
Comment (youknowriad) [inline packages/editor/src/components/url-input/index.js]: I understand the logic, but I'd prefer if we find another way as this seems fragile. It used to work properly before without it (even with two popovers) so curious to know why it regressed.

File: packages/editor/src/components/url-input/index.js [hunk 2]
```diff
@@ -159,22 +169,33 @@ class URLInput extends Component {
...
-	selectLink( post ) {
+	selectLink( post, event ) {
 		this.props.onChange( post.url, post );
 		this.setState( {
 			selectedSuggestion: null,
 			showSuggestions: false,
 		} );
+
+		// Announce a link has been selected when tabbing away from the input field.
+		if ( event.keyCode === TAB ) {
```
Comment (youknowriad) [inline packages/editor/src/components/url-input/index.js]: Why should we only announce this on "tab" only? I prefer to do it consistently if possible.

File: packages/editor/src/components/url-input/index.js (hunk 1 above)
Comment (afercia) [reply packages/editor/src/components/url-input/index.js]: Yeah I agree it's fragile, that's why I've mentioned that maybe the popover component should handle these nesting cases internally. It happened to work before just by coincidence 🙂 If you look at the screenshot posted above, the suggestions list is perfectly centered to the first popover _highlighted area_. The ellipsis button is now outside of the first popover and doesn't contribute to determine its width, so the position calculation of the _second_ popover takes into consideration only the highlighted area. If you move the ellipsis button back inside the first popover, then the position will work correctly (but just by coincidence).

File: packages/editor/src/components/url-input/index.js (hunk 2 above)
Comment (afercia) [reply packages/editor/src/components/url-input/index.js]: Because when pressing Enter, the form is submitted and the link gets _inserted_. There's another speak message for that, see  "Link inserted" in `submitLink()` in `LinkContainer`.

Comment (youknowriad) [reply packages/editor/src/components/url-input/index.js]: Gotcha, instead of passing the event object here, I'd personally just move this to the switch/case above

**Comment by afercia** [comment]
Thanks @youknowriad 
- tried to use the TAB / ENTER cases and dry up a bit to avoid duplicated code
- when a suggested item _gets clicked with the mouse_, focus must be returned on the search field: using a new `handleOnClick()` for this

**Review by youknowriad — APPROVED** [review]
Nice, Thanks for the improvements.

**Review by youknowriad — CHANGES_REQUESTED** [review]
Actually, this broken the url input for the button block.

**Comment by afercia** [comment]
Ah, right. I see the custom `getAnchorRect()` doesn't work on the Button block because it gets the wrong parent and needs the default behavior instead.
Not sure I see other differences from master (or even previous versions), am I missing something? Aside: the interaction on the Button block would need some work.

**Comment by afercia** [comment]
I've reverted the custom `getAnchorRect()` as that's not a proper solution. However, the position thing is still a problem, I guess it would be better addressed separately. Also, I'd tend to think the URL input / popover implementation needs some refinement, looking at it holistically. FOr example, I'm not sure I understand why the Button block URL input doesn't have a "Link Settings" button: as a used, I can I set that link to open in a new tab?

screen shot 2018-10-22 at 08 31 05

**Comment by youknowriad** [comment]
Can you rebase the branch so we can get it in.

**Comment by afercia** [comment]
Rebased. Thank you.

**Comment by talldan** [comment]
I had a test - looks good. A simple solution to the positioning of the autcomplete suggestions might be to make it a child of the editor-url div:
screen shot 2018-11-08 at 4 46 33 pm

Will need to make it a bit narrower, but I think it's still wide enough to cover most post titles.

I think we should tackle that in a separate PR though.

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Comment by afercia** [comment]
> I think we should tackle that in a separate PR though.

@talldan has a PR been created? I've just noticed one more issue: as the ellipsis button is now placed outside, pressing Escape when it's focused or when its expanded panel is focused, doesn't close the link popover.

**Comment by talldan** [comment]
Hi @afercia - here's a PR tackling that:
https://github.com/WordPress/gutenberg/pull/11803

I'll also see if I can solve the escape key issue now. (apologies for these regressions)

**Comment by aduth** [comment]
>makes the TAB key populate the search field with the highlighted link URL [...] adds a speak message to announce the link has been selected, matches the current interaction in the Classic Editor

Was it intentional that we only announce that the link has been selected when explicitly pressing the <kbd>Tab</kbd> key, and not when selecting the link via <kbd>Enter</kbd> ? I could see an argument being made that since it might not be expected that the link is selected when the focus shifts by pressing <kbd>Tab</kbd>, the spoken message helps inform the user about that effect. But if it's unexpected such that it needs to be explained, then it makes me wonder whether it should actually happen at all.

In other words, would you consider either of the following as a reasonable change:

1. Always announce that the link is selected, regardless of whether it occurs by <kbd>Tab</kbd> or by <kbd>Enter</kbd>
2. Or, remove the behavior where a link is selected as a result of pressing <kbd>Tab</kbd>

**Comment by afercia** [comment]
Generally, the `Tab` key should be reserved for keyboard navigation so the most important thing is that it doesn't _insert_ a link.

If I recall correctly, the current behavior is meant to reproduce how `wplink` always worked in WordPress. That's the behavior keyboard and assistive technology users are used to.

In a way, that's not totally unexpected. Some browsers have the same behavior on some native HTML elements. For example on a `<select>` element on macOS:
- Firefox: pressing `Tab` selects the highlighted option
- Webkit (Chrome, Safari, Opera): pressing `Tab` doesn't do anything

Worth noting that when pressing Enter, `wplink` does announce `Link inserted.` so I'd vote for option 1

> Always announce that the link is selected, regardless of whether it occurs by Tab or by Enter
